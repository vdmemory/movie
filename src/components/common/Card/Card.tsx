import React from "react";

import noposter from "./noposter.png"
import styles from "./Card.module.scss";

type Props = {
    title: string;
    poster: string;
    year: string;
    id: string;
    onClick(id: string): void;
};

const Card: React.FC<Props> = ({ title, poster, onClick, id, year }) => {
    const handleClick = () => {
        onClick(id);
    };

    return (
        <div className={styles.movie}>
            <figure onClick={handleClick}>
                <img alt={title} src={poster !== "N/A" ? poster : noposter} />
                <h2>{`${title} (${year})`}</h2>
            </figure>
        </div>
    );
};

export default Card;
