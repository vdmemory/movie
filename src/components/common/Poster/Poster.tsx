import React from "react";

import noposter from "./noposter.png";

type Props = {
    title: string;
    poster: string;
};

const Card: React.FC<Props> = ({ title, poster }) => {
    return <img alt={title} src={poster !== "N/A" ? poster : noposter} />;
};

export default Card;
