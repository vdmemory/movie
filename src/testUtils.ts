export const deepEqual = (a: any, b: any) => {
    if (a === b) {
        return true;
    }

    if (
        a == null ||
        typeof a != "object" ||
        b == null ||
        typeof b != "object"
    ) {
        return false;
    }

    var propertiesInA = 0,
        propertiesInB = 0;
    for (var property in a) {
        propertiesInA += 1;
    }
    for (var property in b) {
        propertiesInB += 1;
        if (!(property in a) || !deepEqual(a[property], b[property])) {
            return false;
        }
    }
    return propertiesInA == propertiesInB;
};
