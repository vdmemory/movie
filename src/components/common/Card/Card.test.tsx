import React from "react";
import { createRenderer } from "react-test-renderer/shallow";

import { deepEqual } from "../../../testUtils";
import Card from "./Card";
import Poster from "../Poster/Poster";

type Props = {
    title: string;
    poster: string;
    year: string;
    id: string;
    onClick(id: string): void;
};

const setup = (props: Props) => {
    const renderer = createRenderer();
    renderer.render(<Card {...props} />);
    const output = renderer.getRenderOutput();
    return { output, props };
};

const propsRender: Props = {
    title: "this title text",
    poster: "https://poster.jpg",
    year: "1990",
    id: "tt55555",
    onClick: jest.fn()
};

describe("Card Component", () => {
    describe("Render component and its components", () => {
        test("Should renders correctly", () => {
            const { output } = setup(propsRender);
            expect(output).toMatchSnapshot();
        });

        test("Should render container and has a movie class", () => {
            const { output } = setup(propsRender);
            expect(output.type).toBe("div");
            expect(output.props.className).toBe("movie");
        });

        test("Should call handleClick on click", () => {
            const { output, props } = setup(propsRender);
            const figure = output.props.children;
            figure.props.onClick();
            expect(props.onClick).toBeCalled();
            expect(figure.type).toBe("figure");;
        });

        test("Should render title and year for poster and should have h2 tag", () => {
            const { output } = setup(propsRender);
            const figure = output.props.children;
            const [, title] = figure.props.children;
            expect(title.type).toBe("h2");
            expect(title.props.children).toBe("this title text (1990)");
        });

        describe("Poster", () => {
            const expectedProps = {
                title: "this title text",
                poster: "https://poster.jpg",
            };
            test("Should render and check props", () => {
                const { output, props } = setup(propsRender);
                const figure = output.props.children;
                const [poster] = figure.props.children;
                expect(poster.type).toBe(Poster);
                expect(deepEqual(poster.props, expectedProps)).toBeTruthy();
            });
        });

        

    });
});
