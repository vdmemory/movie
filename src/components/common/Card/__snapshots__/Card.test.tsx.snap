// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component Render component and its components Should renders correctly 1`] = `
<div
  className="movie"
>
  <figure
    onClick={[Function]}
  >
    <Card
      poster="https://poster.jpg"
      title="this title text"
    />
    <h2>
      this title text (1990)
    </h2>
  </figure>
</div>
`;
